### PR TITLE
trigger support

### PR DESCRIPTION
**Pull request by katiecai**
Add trigger binding support for RabbitMQ

### CODE REVIEW

File: src/Listeners/RabbitMQListener.cs [hunk 1]
```diff
@@ -0,0 +1,118 @@
...
+using System.Threading;
+using System.Threading.Tasks;
+using Microsoft.Azure.WebJobs.Host.Executors;
+using Microsoft.Azure.WebJobs.Host.Listeners;
+using RabbitMQ.Client;
+using RabbitMQ.Client.Events;
+
+namespace Microsoft.Azure.WebJobs.Extensions.RabbitMQ.Listeners
+{
+    internal sealed class RabbitMQListener : IListener
```
[the remark this answers is not in the document]
Comment (pragnagopa) [reply src/Listeners/RabbitMQListener.cs]: Also, move RabbitMQListener into Trigger folder

File: src/Trigger/RabbitMQTriggerAttributeBindingProvider.cs [hunk 2]
```diff
@@ -0,0 +1,84 @@
...
+            ushort batchNumber = attribute.BatchNumber;
+
+            IRabbitMQService service = _provider.GetService(hostname, queueName);
+
+            return new RabbitMQTriggerBinding(service, hostname, queueName, batchNumber);
+        }
+
+        private string Resolve(string name)
+        {
+            if (_nameResolver == null)
```
Comment (pragnagopa) [inline src/Trigger/RabbitMQTriggerAttributeBindingProvider.cs]: shorten _nameResolver.ResolveWholeString(name) ?? name

File: src/Trigger/RabbitMQTriggerAttributeBindingProvider.cs [hunk 3]
```diff
@@ -0,0 +1,84 @@
...
+            if (attribute == null)
+            {
+                return null;
+            }
+
+            string queueName = null;
+
+            if (attribute.QueueName != null)
+            {
+                queueName = this.Resolve(attribute.QueueName);
```
Comment (pragnagopa) [inline src/Trigger/RabbitMQTriggerAttributeBindingProvider.cs]: nit: remove this

File: src/Trigger/RabbitMQTriggerAttributeBindingProvider.cs [hunk 4]
```diff
@@ -0,0 +1,84 @@
...
+            }
+
+            string hostname = null;
+
+            if (attribute.Hostname != null)
```
Comment (pragnagopa) [inline src/Trigger/RabbitMQTriggerAttributeBindingProvider.cs]: Do not have to check for null here as RabbitMQService should check for null before creating a connection.

File: src/Trigger/RabbitMQTriggerBinding.cs [hunk 5]
```diff
@@ -0,0 +1,117 @@
...
+            }
+
+            return Task.FromResult<IListener>(new RabbitMQListener(context.Executor, _service, _queueName, _batchNumber));
+        }
+
+        public ParameterDescriptor ToParameterDescriptor()
+        {
+            return new RabbitMQTriggerParameterDescriptor
+            {
+                Hostname = _hostname,
```
Comment (pragnagopa) [inline src/Trigger/RabbitMQTriggerBinding.cs]: nit: use pascalCasing

File: src/Trigger/RabbitMQTriggerParameterDescriptor.cs [hunk 6]
```diff
@@ -0,0 +1,14 @@
...
+
+using Microsoft.Azure.WebJobs.Host.Protocols;
+
+namespace Microsoft.Azure.WebJobs.Extensions.RabbitMQ.Trigger
+{
+    internal class RabbitMQTriggerParameterDescriptor : TriggerParameterDescriptor
+    {
+        public string Hostname { get; set; }
+
+        public string QueueName { get; set; }
```
Comment (pragnagopa) [inline src/Trigger/RabbitMQTriggerParameterDescriptor.cs]: Add trigger reason

File: test/WebJobs.Extensions.RabbitMQ.Samples/Program.cs [hunk 7]
```diff
@@ -14,8 +14,13 @@ public static class Program
...
         public static async Task Main(string[] args)
         {
+            string filesTestPath = @"c:\temp\files";
+
+            // Add or remove types from this list to choose which functions will
+            // be indexed by the JobHost.
+            // To run some of the other samples included, add their types to this list
             var typeLocator = new SamplesTypeLocator(
-                typeof(RabbitMQSamples));
+                typeof(RabbitMQTriggerSamples));
```
Comment (pragnagopa) [inline test/WebJobs.Extensions.RabbitMQ.Samples/Program.cs]: combine RabbitMQSamples and RabbitMQTriggerSamples

File: test/WebJobs.Extensions.RabbitMQ.Samples/RabbitMQTriggerSamples.cs [hunk 8]
```diff
@@ -0,0 +1,65 @@
...
+            [RabbitMQTrigger("localhost", "queue")] string message,
+            string consumerTag,
+            ILogger logger
+        )
+        {
+            logger.LogInformation($"RabbitMQ queue trigger function processed message consumer tag: {consumerTag}");
+            logger.LogInformation($"RabbitMQ queue trigger function processed message: {message}");
+        }
+
+        public static void RabbitMQTrigger_BasicDeliverEventArgs(
```
Comment (pragnagopa) [inline test/WebJobs.Extensions.RabbitMQ.Samples/RabbitMQTriggerSamples.cs]: Add another sample that uses RabbitMQTrigger and RabbitMQ output binding- listening on one queue and writing to a different queue

File: test/WebJobs.Extensions.RabbitMQ.Tests/RabbitMQConfigurationTests.cs [hunk 9]
```diff
@@ -10,6 +10,10 @@
 using Newtonsoft.Json;
 using RabbitMQ.Client;
 using Xunit;
+using System;
+using System.Collections.Generic;
```
Comment (pragnagopa) [inline test/WebJobs.Extensions.RabbitMQ.Tests/RabbitMQConfigurationTests.cs]: nit:sort using statements giving precendence to system.

File: test/WebJobs.Extensions.RabbitMQ.Tests/RabbitMQConfigurationTests.cs [hunk 10]
```diff
@@ -119,12 +123,83 @@ public void Converts_String_Correctly()
...
+
+        [Fact]
+        public void Null_Context_Throws_Error()
+        {
+            var mockProvider = new Mock<RabbitMQTriggerAttributeBindingProvider>();
+            Assert.ThrowsAsync<ArgumentNullException>(() => mockProvider.Object.TryCreateAsync(null));
+        }
+
+        [Fact]
+        public void Creates_Contract_Correctly()
```
Comment (pragnagopa) [inline test/WebJobs.Extensions.RabbitMQ.Tests/RabbitMQConfigurationTests.cs]: nit: raname test to Verify_BindingDataContract_Types

File: test/WebJobs.Extensions.RabbitMQ.Tests/RabbitMQConfigurationTests.cs [hunk 11]
```diff
@@ -119,12 +123,83 @@ public void Converts_String_Correctly()
...
+            var actualContract = RabbitMQTriggerBinding.CreateBindingDataContract(); 
+
+            foreach (KeyValuePair<string, Type> item in actualContract)
+            {
+                Assert.Equal(expectedContract[item.Key], item.Value);
+            }
+        }
+
+        [Fact]
+        public void Correctly_Populates_Binding_Data()
```
Comment (pragnagopa) [inline test/WebJobs.Extensions.RabbitMQ.Tests/RabbitMQConfigurationTests.cs]: nit: raname test to Verify_BindingDataContract_Values

File: src/Config/RabbitMQExtensionConfigProvider.cs [hunk 12]
```diff
@@ -2,27 +2,36 @@
...
+using RabbitMQ.Client.Events;
 
 namespace Microsoft.Azure.WebJobs.Extensions.RabbitMQ
 {
     [Extension("RabbitMQ")]
     internal class RabbitMQExtensionConfigProvider : IExtensionConfigProvider
     {
         private readonly IOptions<RabbitMQOptions> _options;
+        private readonly IConfiguration _configuration;
+        private readonly INameResolver _nameResolver;
```
Comment (pragnagopa) [inline src/Config/RabbitMQExtensionConfigProvider.cs]: this is not assigned?

File: src/Config/RabbitMQExtensionConfigProvider.cs [hunk 13]
```diff
@@ -2,27 +2,36 @@
...
         private readonly IRabbitMQServiceFactory _rabbitMQServiceFactory;
         private readonly ILoggerFactory _loggerFactory;
+        private ILogger _logger;
 
-        public RabbitMQExtensionConfigProvider(IOptions<RabbitMQOptions> options, IRabbitMQServiceFactory rabbitMQServiceFactory, ILoggerFactory loggerFactory)
+        public RabbitMQExtensionConfigProvider(IOptions<RabbitMQOptions> options, IConfiguration configuration, IRabbitMQServiceFactory rabbitMQServiceFactory, ILoggerFactory loggerFactory)
         {
```
Comment (pragnagopa) [inline src/Config/RabbitMQExtensionConfigProvider.cs]: Should it  be INameResolver nameResolver instead of IConfiguration configuration?

File: test/WebJobs.Extensions.RabbitMQ.Tests/RabbitMQConfigurationTests.cs [hunk 14]
```diff
@@ -1,33 +1,32 @@
...
-        private static readonly IConfiguration _emptyConfig = new ConfigurationBuilder().Build();
-
         [Fact]
         public void Creates_Context_Correctly()
         {
             var options = new RabbitMQOptions { Hostname = "localhost", QueueName = "hello" };
             var loggerFactory = new LoggerFactory();
             var mockServiceFactory = new Mock<IRabbitMQServiceFactory>();
-            var config = new RabbitMQExtensionConfigProvider(new OptionsWrapper<RabbitMQOptions>(options), _emptyConfig, mockServiceFactory.Object, (ILoggerFactory)loggerFactory);
+            var mockNameResolver = new Mock<INameResolver>();
```
Comment (katiecai) [inline test/WebJobs.Extensions.RabbitMQ.Tests/RabbitMQConfigurationTests.cs]: Mocked the name resolver for now since other extensions use the common test framework

File: src/Config/RabbitMQExtensionConfigProvider.cs [hunk 15]
```diff
@@ -152,11 +147,15 @@ public byte[] Convert(T input)
 
         internal class EventArgsToPocoConverter<T> : IConverter<BasicDeliverEventArgs, T>
         {
+            private readonly LoggerFactory _loggerFactory = new LoggerFactory();
```
Comment (katiecai) [inline src/Config/RabbitMQExtensionConfigProvider.cs]: Created member variables for logging for internal class, but is there a way to pass in the logger factory/logger from the actual config provider into the converter?

File: src/Trigger/RabbitMQTriggerAttributeBindingProvider.cs [hunk 16]
```diff
@@ -17,18 +16,15 @@ internal class RabbitMQTriggerAttributeBindingProvider : ITriggerBindingProvider
     {
         private readonly INameResolver _nameResolver;
         private readonly ILogger _logger;
```
Comment (katiecai) [inline src/Trigger/RabbitMQTriggerAttributeBindingProvider.cs]: _logger and _options aren't currently read (and I get this warning on VS) but could be used in the future. Wasn't sure if I should keep them both for now

File: src/Trigger/RabbitMQTriggerAttributeBindingProvider.cs [hunk 17]
```diff
@@ -0,0 +1,84 @@
...
+            ILoggerFactory loggerFactory)
+        {
+            _config = configuration;
+            _nameResolver = nameResolver;
+            _options = options;
+            _provider = provider;
+            _logger = loggerFactory?.CreateLogger(LogCategories.CreateTriggerCategory("RabbitMQ"));
+        }
+
+        public async Task<ITriggerBinding> TryCreateAsync(TriggerBindingProviderContext context)
```
Comment (katiecai) [inline src/Trigger/RabbitMQTriggerAttributeBindingProvider.cs]: This never awaits because I never make any async calls in this function, but follows the programming model of other extensions. Should I still keep it as async?

File: src/Config/BasicDeliverEventArgsToPocoConverter.cs [hunk 18]
```diff
@@ -0,0 +1,42 @@
...
+            {
+                jsonObj = JToken.Parse(body);
+            }
+            catch (FormatException fex)
+            {
+                throw new FormatException(fex.ToString());
+            }
+            catch (Exception ex)
+            {
+                _logger.LogError(ex.ToString());
```
Comment (pragnagopa) [inline src/Config/BasicDeliverEventArgsToPocoConverter.cs]: Log exception object.  Change to `logger?.LogError(ex, $"Failed converting BasicDeliverEventArgs body to To Poco")`

File: src/Config/BasicDeliverEventArgsToPocoConverter.cs [hunk 19]
```diff
@@ -0,0 +1,42 @@
...
+        public T Convert(BasicDeliverEventArgs arg)
+        {
+            string body = Encoding.UTF8.GetString(arg.Body);
+            JToken jsonObj = null;
+
+            try
+            {
+                jsonObj = JToken.Parse(body);
+            }
+            catch (FormatException fex)
```
Comment (pragnagopa) [inline src/Config/BasicDeliverEventArgsToPocoConverter.cs]: Delete `catch (FormatException fex)`

File: src/Config/BasicDeliverEventArgsToPocoConverter.cs [hunk 20]
```diff
@@ -0,0 +1,42 @@
...
+            {
+                throw new FormatException(fex.ToString());
+            }
+            catch (Exception ex)
+            {
+                _logger.LogError(ex.ToString());
+            }
+
+            return jsonObj.ToObject<T>();
```
Comment (pragnagopa) [inline src/Config/BasicDeliverEventArgsToPocoConverter.cs]: perhaps return null if the conversion fails? Also, add tests for successful conversion and failure case

File: test/WebJobs.Extensions.RabbitMQ.Tests/RabbitMQOutputBindingTests.cs [hunk 21]
```diff
@@ -63,63 +62,35 @@ public void Handles_Null_Attributes_And_Options(string attrHostname, string attr
...
-            var collector = new RabbitMQAsyncCollector(context, logger);
-
-            byte[] body = Encoding.UTF8.GetBytes("hi");
-            await collector.AddAsync(body);
-
-            mockBatch.Verify(m => m.Add(It.IsAny<string>(), It.IsAny<string>(), It.IsAny<bool>(), It.IsAny<IBasicProperties>(), body), Times.Exactly(1));
-        }
-
         [Fact]
         public void Converts_String_Correctly()
```
Comment (pragnagopa) [inline test/WebJobs.Extensions.RabbitMQ.Tests/RabbitMQOutputBindingTests.cs]: Move this to its own  test file. Also add a test where conversion fails and throws exception. Add test for null

File: src/Config/RabbitMQExtensionConfigProvider.cs [hunk 22]
```diff
@@ -38,18 +41,30 @@ public void Initialize(ExtensionConfigContext context)
...
+                .AddConverter<BasicDeliverEventArgs, DirectInvokeString>((args) => new DirectInvokeString(null));
+
+            // Convert BasicDeliverEventArgs --> string-- > JSON-- > POCO
+            triggerRule.AddOpenConverter<BasicDeliverEventArgs, OpenType.Poco>(typeof(BasicDeliverEventArgsToPocoConverter<>), _logger);
         }
 
         public void ValidateBinding(RabbitMQAttribute attribute, Type type)
         {
-            string hostname = Utility.FirstOrDefault(attribute.Hostname, _options.Value.Hostname);
+            string hostname = Utility.FirstOrDefault(attribute.HostName, _options.Value.Hostname);
```
Comment (pragnagopa) [inline src/Config/RabbitMQExtensionConfigProvider.cs]: nit: hostName, queueName

File: test/WebJobs.Extensions.RabbitMQ.Tests/RabbitMQOutputBindingTests.cs [hunk 23]
```diff
@@ -1,36 +1,35 @@
...
 using Moq;
 using Newtonsoft.Json;
-using RabbitMQ.Client;
 using Xunit;
 
-namespace Microsoft.Azure.WebJobs.Extensions.RabbitMQ.Tests
+namespace WebJobs.Extensions.RabbitMQ.Tests
 {
-    public class RabbitMQConfigurationTests
+    public class RabbitMQOutputBindingTests
```
Comment (pragnagopa) [inline test/WebJobs.Extensions.RabbitMQ.Tests/RabbitMQOutputBindingTests.cs]: Should this be RabbitMQExtensionConfigProviderTests?

File: test/WebJobs.Extensions.RabbitMQ.Tests/RabbitMQOutputBindingTests.cs [hunk 24]
```diff
@@ -1,36 +1,35 @@
...
         [Fact]
         public void Creates_Context_Correctly()
         {
             var options = new RabbitMQOptions { Hostname = "localhost", QueueName = "hello" };
             var loggerFactory = new LoggerFactory();
             var mockServiceFactory = new Mock<IRabbitMQServiceFactory>();
-            var config = new RabbitMQExtensionConfigProvider(new OptionsWrapper<RabbitMQOptions>(options), mockServiceFactory.Object, (ILoggerFactory)loggerFactory);
-            var attribute = new RabbitMQAttribute { Hostname = "localhost", QueueName = "queue" };
+            var mockNameResolver = new Mock<INameResolver>();
+            var config = new RabbitMQExtensionConfigProvider(new OptionsWrapper<RabbitMQOptions>(options), mockNameResolver.Object, mockServiceFactory.Object, (ILoggerFactory)loggerFactory);
```
Comment (pragnagopa) [inline test/WebJobs.Extensions.RabbitMQ.Tests/RabbitMQOutputBindingTests.cs]: test with different host names i.e do not use localhost for options and attribute

File: src/Services/RabbitMQService.cs [hunk 25]
```diff
@@ -1,30 +1,44 @@
...
         private IBasicPublishBatch _batch;
 
-        public RabbitMQService(string hostname, string queuename)
+        public RabbitMQService(string hostName, string queueName)
         {
-            _batch = CreateBatch(hostname, queuename);
+            string host = hostName ?? throw new ArgumentNullException(nameof(hostName));
+            string queue = queueName ?? throw new ArgumentNullException(nameof(queueName));
+            CreateChannel(host);
+            CreateBatch(queue);
```
Comment (pragnagopa) [inline src/Services/RabbitMQService.cs]: Add private variables for hostName and queueName. Move logic from CreateChannel and CreateBatch to constructor as those methods are not being used any where else

File: src/Trigger/RabbitMQListener.cs [hunk 26]
```diff
@@ -0,0 +1,112 @@
...
+            if (_started)
+            {
+                throw new InvalidOperationException("The listener has already been started.");
+            }
+
+            _channel = _service.GetChannel();
+            _channel.BasicQos(0, _batchNumber, false);
+            _consumer = new EventingBasicConsumer(_channel);
+
+            // TODO:
```
Comment (pragnagopa) [inline src/Trigger/RabbitMQListener.cs]: Please open an issue for this TODO and you can delete the comments from here. Add details to the issue and we will address it when needed

File: src/Trigger/RabbitMQTriggerAttributeBindingProvider.cs (hunk 16 above)
Comment (pragnagopa) [reply src/Trigger/RabbitMQTriggerAttributeBindingProvider.cs]: if they are not used, please delete. That will clean up constructor as well.

File: src/Trigger/RabbitMQTriggerParameterDescriptor.cs [hunk 27]
```diff
@@ -1,14 +1,20 @@
...
 {
     internal class RabbitMQTriggerParameterDescriptor : TriggerParameterDescriptor
     {
         public string Hostname { get; set; }
 
         public string QueueName { get; set; }
+
+        public override string GetTriggerReason(IDictionary<string, string> arguments)
+        {
+            return string.Format("RabbitMQ message detected from queue: {0}", this.QueueName);
```
Comment (pragnagopa) [inline src/Trigger/RabbitMQTriggerParameterDescriptor.cs]: add date time: `new RabbitMQ message detected on queue: {0} at {1}`

File: test/WebJobs.Extensions.RabbitMQ.Tests/RabbitMQTriggerBindingTests.cs [hunk 28]
```diff
@@ -0,0 +1,96 @@
...
+using RabbitMQ.Client;
+using RabbitMQ.Client.Events;
+using Xunit;
+
+namespace WebJobs.Extensions.RabbitMQ.Tests
+{
+    public class RabbitMQTriggerBindingTests
+    {
+        [Fact]
+        public void Converts_to_Poco()
```
Comment (pragnagopa) [inline test/WebJobs.Extensions.RabbitMQ.Tests/RabbitMQTriggerBindingTests.cs]: Move this test to its own file : BasicDeliverEventArgsToPocoConverterTests
